index.ts

```diff
@@ -0,0 +1,552 @@
+// Exported native API for @smogon/data
+//
+// DISCLAIMER:
+//
+//   Not meant to replace Dex/BattleDex: consumers can delegate to this native API
+//   and put their own layer on top which can extend or alter the functionality.
+//
+// CONSIDERATIONS:
+//
+//    - Approach: async APIs are moved to top level so that data load can be done
+//      once and then the rest of the APIs aren't cluttered with `await` everywhere.
+//      Furthermore, individual pieces can be loaded (just `Abilities` vs. all data
+//      types) if desired.
+//    - `getByID` APIs are important for removing the `toID` call in `get` methods
+//      for when we already know through typechecking that we have an ID. String
+//      getters have been removed - callees can manually called `toID` first if they
+//      dont have an ID (which should be very rare unless its human-input).
+//    - Descriptions are UI only and can be large so aren't included in the Dex by
+//      default, but they may be loaded in bulk asynchronously for the desired
+//      languages.
+//    - Similarly, SpeciesDetails which are purely cosmetic or used by the Validator
+//      or for non-competitive reasons are offloaded. eg. learnset details are very
+//      large and not needed in many circumstances.
+//    - All other APIs are loaded together for a gen (+ mod) - it's rare a consumer
+//      cares *just* about Abilities or *just* Moves, and if they do and are in an
+//      environment where loading more is a concern they can implement their own
+//      splitting on top of the raw data files.
+//    - The aliases.json data is used internally (to implement getByID's) but is
+//      not exported.
+//    - The individual APIs (eg. Abilities) are not just Maps, as their lookup function
+//      is more sophisticated (handles aliases, caching, etc).
+//    - Internally, references can be resolved lazily through Javascript getters (though
+//      need to use `Object.defineProperty(Foo.prototype, ...)` to be able to make
+//      enumerable) to avoid object instantiation overhead and avoid creating objects for
+//      fields which are never read.
```

Comment:
Maybe you didn't mean this to be the actual implementation strategy, but I would imagine we'd define `Foo` as an ES6 class with `get property() { ... }` methods?

Reply:
get property methods are not enumerable

Reply:
Getters aren't polyfillable, have poor performance, and are too magical for my tastes. :( Would rather use an explicit getter method like `getProperty()`.

Reply:
![results](https://user-images.githubusercontent.com/117249/67047121-8b328900-f0e6-11e9-90a7-24a4a04d1637.png)

Is what https://jsperf.com/accessors-vs-functions/1 results look lie on my box - a 4% isnt a super big hit. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get seems to imply `get` is available back to IE9...

Reply:
Huh, I remember them being a pretty big hit last time I looked into it; I guess they got better. Still, readability and polyfillability matter to me.

Reply:
Can we just drop the unnecessary IE7 support already? IE9 is still plenty old. :grinning: 

As for readability - can you elaborate on what the concerns are here? What makes it less readable?

Reply:
I would prefer for reading and writing properties not to secretly execute code, mostly.

Reply:
Also: it's nice to see at a glance what is state and what is a computed property, and if a computed property ends in `()` that makes the difference obvious to me.

Reply:
secretly executing code and looking exactly like a property is a feature - its an implementation detail whether something is computed or not (code to build the object will run on the first access and then be cached, at which points its the same as a lookup anyway)

Reply:
I'm aware it's a feature, it's just not desirable to me. Knowing what is and isn't state is useful to me.

Reply:
it is state, it just hasnt been resolved yet. if we resolve all the references on creation we just end up slowing down creation and creating a bunch of objects we might never use. if we lazy load and user getters we can still show the same API but change the performance tradeoffs